crates/turbopack-ecmascript/src/references/esm/meta.rs

```diff
@@ -0,0 +1,66 @@
+use anyhow::Result;
+use swc_core::{
+    common::DUMMY_SP,
+    ecma::ast::{Expr, Ident},
+    quote,
+};
+use turbo_tasks_fs::FileSystemPathVc;
+use turbopack_core::chunk::ChunkingContextVc;
+
+use crate::{
+    code_gen::{CodeGenerateable, CodeGenerateableVc, CodeGeneration, CodeGenerationVc},
+    create_visitor, magic_identifier,
+    references::{esm::base::insert_hoisted_stmt, AstPathVc},
+};
+
+#[turbo_tasks::value(shared)]
+#[derive(Hash, Debug)]
+pub struct ImportMetaRef {
+    path: FileSystemPathVc,
+    initialize: bool,
+    ast_path: AstPathVc,
+}
+
+#[turbo_tasks::value_impl]
+impl ImportMetaRefVc {
+    #[turbo_tasks::function]
+    pub fn new(path: FileSystemPathVc, initialize: bool, ast_path: AstPathVc) -> Self {
+        ImportMetaRef {
+            path,
+            initialize,
+            ast_path,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl CodeGenerateable for ImportMetaRef {
+    #[turbo_tasks::function]
+    async fn code_generation(&self, _context: ChunkingContextVc) -> Result<CodeGenerationVc> {
+        // TODO: should only be done in ESM
+        let ast_path = &self.ast_path.await?;
+        let mut visitors = vec![create_visitor!(ast_path, visit_mut_expr(expr: &mut Expr) {
+            let id = Ident::new(magic_identifier::encode("import.meta").into(), DUMMY_SP);
+            *expr = Expr::Ident(id);
+        })];
+
+        // There can be many references to import.meta, and they appear at nesting in
+        // the file. The first reference is responsible for injecting the
+        // module-level variable to hold the mutable "meta object".
+        if self.initialize {
+            let path = format!("/{}", self.path.await?.path);
+            visitors.push(create_visitor!(visit_mut_program(program: &mut Program) {
+                let name = Ident::new(magic_identifier::encode("import.meta").into(), DUMMY_SP);
+                let meta = quote!(
+                    "const $name = { url: new URL($url, location.href).href };" as Stmt,
```

Comment:
What's the plan for node.js execution? Here `location` doesn't exist.

webpack defines a baseURI here which differs depending on environment. It always points to the output directory.

* node.js: `require("url").pathToFileURL(__dirname + "...")`
* ESM: `new URL("...", import.meta.url)`
* web: `document.baseURI || self.location.href`

Reply:
I've updated this to use [as_abs_path](https://github.com/vercel/turbo/blob/2ee0b494611cc3d2a8e400d541174de4a7ef8a8f/crates/turbopack-ecmascript/src/references/mod.rs#L1330-L1332), which is what `__dirname` and `__filename` use to get an "absolute" path. We might want to update this to generate a real sys path in the future, but at least this makes it consistent between CJS and ESM.